### PR TITLE
Updates Readme sections for `only` and `ignore`

### LINKED ISSUE #1240: Ignore or Skip attributes dynamically is not working
- [x] This is not a usage question, this is a bug report
- [x] This bug can be reproduced with the script I provide below
- [x] This bug can be reproduced in the latest release of the `paper_trail` gem

```ruby
# frozen_string_literal: true

# Use this template to report PaperTrail bugs.
# Please include only the minimum code necessary to reproduce your issue.
require "bundler/inline"

# STEP ONE: What versions are you using?
gem 'paper_trail', '10.3.1'
gem 'rails', '5.2.3'

ActiveRecord::Schema.define do
  # STEP TWO: Define your tables here.
  create_table :users, force: true do |t|
    t.text :first_name, null: false
    t.timestamps null: false
  end

  create_table :versions do |t|
    t.string :item_type, null: false
    t.integer :item_id, null: false
    t.string :event, null: false
    t.string :whodunnit
    t.text :object, limit: 1_073_741_823
    t.text :object_changes, limit: 1_073_741_823
    t.datetime :created_at
  end
  add_index :versions, %i[item_type item_id]
end
ActiveRecord::Base.logger = Logger.new(STDOUT)
require "paper_trail"

# STEP FOUR: Define your AR models here.
class User < ActiveRecord::Base
  has_paper_trail on: %i[update destroy], ignore: { first_name: Proc.new { |obj|  true } }
end

# STEP FIVE: Please write a test that demonstrates your issue.
class BugTest < ActiveSupport::TestCase
  def test_1
    User.create(first_name: "Jane")
    assert_difference(-> { PaperTrail::Version.count },  0) {
      User.update(first_name: "Jane 2")
    }
  end
end

# STEP SIX: Run this script using `ruby my_bug_report.rb`
```

### PR DESCRIPTION
**Pull request by tlynam**
Thank you for your contribution!

Check the following boxes:

- [x] Wrote [good commit messages][1].
- [x] Feature branch is up-to-date with `master` (if not - rebase it).
- [x] Squashed related commits together.
- [ ] Added tests.
- [ ] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new 
  code introduces user-observable changes.
- [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by tlynam** [comment]
Hi @jaredbeck, hope you're well.  It looks like `ignore` was only working for arrays.  Ignore was working for hashes when calculating `changed_and_not_ignored`.  Can we reuse this code for both cases?  Once it's flushed out, I can add tests and update the changelog.

**Comment by jaredbeck** [comment]
I couldn't find documentation re: hashes in the `:ignore` array. Thankfully, we do have tests in `spec/models/article_spec.rb`. Based on my review of those tests, and `events/base.rb`, it seems like https://github.com/paper-trail-gem/paper_trail/issues/1240 is not a bug report, but rather, is proposing a new feature.

Currently, we allow the `:ignore` array to *contain* hashes. By comparison, https://github.com/paper-trail-gem/paper_trail/issues/1240 proposes that `:ignore` *be* a hash. Therefore, https://github.com/paper-trail-gem/paper_trail/issues/1240 proposes a new feature.

I'm hesitant to increase the complexity of our public API in this way.

**Comment by jaredbeck** [comment]
Frankly, the current API, which allows the `:ignore` array to *contain* hashes, seems oddly complicated to me. I'd be open to simplifying it in our next major version. Perhaps `:ignore` should only accept an array of symbols, and a new option `:ignore_if` could take a `Proc`.

**Comment by tlynam** [comment]
Thank you @jaredbeck!  Can you please help me understand this part of the documentation?  To me, it reads like `ignore` can be a hash?

```
The :ignore and :only options can also accept Hash arguments.

class Article < ActiveRecord::Base
  has_paper_trail only: { title: Proc.new { |obj| !obj.title.blank? } }
end
```

**Comment by jaredbeck** [comment]
Ah, good find. But, looking at `Events::Base#notably_changed`, I think that documentation is incorrect?

https://github.com/paper-trail-gem/paper_trail/blob/21ead888296f022ad63d907b4b15ea598596d1e6/lib/paper_trail/events/base.rb#L221-L235

As with `:ignore`, I think that `:only` can *contain* hashes, but cannot *be* a hash.

So, now I'm not sure whether the docs are wrong or the implementation is wrong 😅 but I think they're *different*. 😓

**Comment by tlynam** [comment]
Ah, thank you!  I see that now.  Okay, would the simplest route be to update the documentation for `ignore` describing that it only works with an array of symbols?  I like your idea of separating `ignore` and `ignore_if`.  Would `ignore_if` be a hash of procs?

**Comment by jaredbeck** [comment]
> .. would the simplest route be to update the documentation ..

Yeah, let's just update the docs for now.

> .. [Should the docs] for `ignore` describing that it only works with an array of symbols?

Yes, but we should also give an example of an array of symbols that also *contains* a hash. See `spec/dummy_app/app/models/article.rb`. We may want to say that we are considering deprecating the later.

> I like your idea of separating `ignore` and `ignore_if`. Would `ignore_if` be a hash of procs?

I was thinking it'd be a single proc. WDYT?

**Comment by tlynam** [comment]
Ah, thank you, I was confused by what you meant by `ignore_if`.

Okay, so the bug report is mentioning that a hash that's inside of an array isn't working as expected for `ignore`.  It looks like it responds to a hash within an array for `changed_and_not_ignored` and was trying to extract and use this same logic for `ignored_attr_has_changed?` as well.  It looks like `ignored_attr_has_changed?` may just work with an array without hashes.  Do you see the same thing?  Is it possible there isn't an existing test for `ignore` with a hash within an array?

**Comment by jaredbeck** [comment]
>  [is there] an existing test for ignore with a hash within an array?

Yes, `spec/models/article_spec.rb`.

I'd suggest that we focus, for now, on adding examples of current behavior to the docs (section 2.c.) and hold off on discussing any changes to the code until that's been done.

File: README.md [hunk 1]
```diff
@@ -428,6 +428,16 @@ a.versions.length                         # 2
...
 
+The `:ignore` option can also accept `Hash` arguments that we are considering deprecating.
+
+```ruby
+class Article < ActiveRecord::Base
+  has_paper_trail ignore: [:title, { color: proc { |obj| obj.color == "Yellow" } }]
+end
+```
+
+If an attribute is ignored and only that attribute was updated, then a version won't be created for `updated_at`.
```
Comment (tlynam) [inline README.md:439]: I didn't see that we document this.  Is it located somewhere else?

**Comment by tlynam** [comment]
Hi @jaredbeck, I hope you're well.

1) Okay, what do you think of these Readme updates saying we support ignore with an array of symbols and hashes with an example?
2) Do we need to document that we don't save a version row for timestamp attributes if the only updated column is ignored?
3) I believe I was able to reproduce the reported bug as a failing test in `spec/models/gadget_spec.rb`.  It's saving a version record for `updated_at` when the updated column is ignored via Hash.  This doesn't occur when the column is ignored via symbol.  I believe the changes in `lib/paper_trail/events/base.rb` addresses it.  I think `changed_notably?` would evaluate `ignored_attr_has_changed?` as false when it should be true and so it would create a version row for `updated_at`.
4) If you see this as a bug, should we address it?

File: README.md [hunk 2]
```diff
@@ -450,7 +460,7 @@ a.versions.length                         # 2
 a.paper_trail.previous_version.content    # nil
 ```
 
-The `:ignore` and `:only` options can also accept `Hash` arguments.
+The `:only` option can also accept `Hash` arguments that we are considering deprecating.
```
Comment (jaredbeck) [inline README.md:463]: I think the example that follows this line is still incorrect. It shows `only: {}`, but should be `only: []`, right?

File: README.md [hunk 3]
```diff
@@ -428,6 +428,16 @@ a.versions.length                         # 2
...
 
+The `:ignore` option can also accept `Hash` arguments that we are considering deprecating.
+
+```ruby
+class Article < ActiveRecord::Base
+  has_paper_trail ignore: [:title, { color: proc { |obj| obj.color == "Yellow" } }]
+end
+```
+
+If an attribute is ignored and only that attribute was updated, then a version won't be created for timestamp attributes.
```
Comment (jaredbeck) [inline README.md:439]: I may be misunderstanding, but I think this has already been said, above.

**Comment by jaredbeck** [comment]
>  Do we need to document that we don't save a version row for timestamp attributes if the only updated column is ignored?

Sure, but can you make it a separate PR please?

>    I believe I was able to reproduce the reported bug as a failing test in spec/models/gadget_spec.rb. It's saving a version record for updated_at when the updated column is ignored via Hash. This doesn't occur when the column is ignored via symbol. I believe the changes in lib/paper_trail/events/base.rb address it. I think changed_notably? would evaluate ignored_attr_has_changed? as false when it should be true and so it would create a version row for updated_at.
>    If you see this as a bug, should we address it?

I may be missing something, but I don't see the need to make any code changes. For now, I'm satisfied with correcting the documentation.

File: README.md (hunk 3 above)
Comment (tlynam) [reply README.md:439]: Thank you, I missed that 👍

File: README.md (hunk 2 above)
Comment (tlynam) [reply README.md:461]: Thank you!  Updated the example 👍

**Comment by tlynam** [comment]
Thank you @jaredbeck, this PR now just updates the Readme.  I created a new PR with a failing test https://github.com/paper-trail-gem/paper_trail/pull/1256

I also created a PR to set the Rubocop gem to a fixed minor version so CI won't install a newer version that starts erroring in CI.  It also updates the Rubocop gems.  https://github.com/paper-trail-gem/paper_trail/pull/1254

**Pull request merged**